Integrate 3scale-authorizer cache

philipgough opened the pull request:
@unleashed - continuing on with the refactor here. Tests won't pass until all updates are done

    authorizer: Introduce package to manage end-to-end authorization
    
    This change introduces a new package with constructs that will
    be capble of carrying out full authorization capabilities between
    an arbitary service and 3scale.
    
    As of this commit, the 'Manager' struct will be capable
    of building out the required clients for 3scale, moving and
    reusing the Builder interface that was created previously.
    
    The 'Manager' now pulls in the previously abstracted system
    caching functionality which can be invoked or otherwise but
    for the caller, it hides the cache and simplifies the interface with
    'Manager'

philipgough committed:
authorizer: Introduce package to manage end-to-end authorization

This change introduces a new package with constructs that will
be capble of carrying out full authorization capabilities between
an arbitary service and 3scale.

As of this commit, the 'Manager' struct will be capable
of building out the required clients for 3scale, moving and
reusing the Builder interface that was created previously.

The 'Manager' now pulls in the previously abstracted system
caching functionality which can be invoked or otherwise but
for the caller, it hides the cache and simplifies the interface with
'Manager'

philipgough committed:
test:authorizer: Unit tests for builder and authorization manager

philipgough committed:
dep: Pull in system cache and latest porta client

PhilipGough commented:
@jmprusi - interested in your feedback here also. This object will act as an "authoriser" and is basically ready for the backend cache to plug into, providing full functionality

philipgough committed:
authorizer: Adds backend types and AuthRep

philipgough committed:
authorizer: Adds backend types and AuthRep

philipgough committed:
test:authorizer - Unit tests for AuthRep and request transformation

philipgough committed:
test:authorizer - Unit tests for AuthRep and request transformation

philipgough committed:
threescale: Update integration and unit tests to support refactor

This change moves away from the previous testing format where we
duplicated the fake http requests in our testing packages and makes use
of the interfaces we created and uses mocks.

This improves the test coverage, but more importantly, cleans up the
tests and makes them more valuable.

This moves out as much testing as possible from the unit tests to the
istio adapter integration testing framework which is more thorough and
tests against a really instance of mixer.

Seperating out the handler allos us to build integration tests with a
variety of inputs as opposed to that hardcoded in the testdata
directory.

philipgough committed:
threescale: Delete replaced constructs

This change delete the previous standalone proxy configuration for
system which will now be provided from an external depedency.

Removes the client builder which will now come from a different package.

Removes the concept of the backend connector which can now be provided
as an interface.

philipgough committed:
test:kubernetes - Updates test to support decoupled handler

philipgough committed:
threescale: Refactor adapter

Restructures the adapter to make use of its provided dependencies
instead of doing the work directly in the handler.

The adapter now is responsible only for request transformation and
validation while the actual work is done by the Authorizer.

philipgough committed:
authorizer: Adds constructor for config and run background processs for refresh on creation

philipgough committed:
cmd:server: Wires together updated authorization code

jmprusi commented on pkg/threescale/authorizer/authorizer.go:
```diff
@@ -0,0 +1,276 @@
+// Authorizer fetches configuration from 3scale and authorizes requests to 3scale
+type Authorizer interface {
+	GetSystemConfiguration(systemURL string, request SystemRequest) (client.ProxyConfig, error)
+	AuthRep(backendURL string, request BackendRequest) (*BackendResponse, error)
```
Could you expose: 
* authorize
* report
* authrep ?

PhilipGough replied:
Yes, will bring them in when we add the backend cache. Thats defiently the idea

unleashed commented on pkg/threescale/authorizer/authorizer.go:
```diff
@@ -0,0 +1,276 @@
+
+func (m Manager) generateSystemCacheKey(systemURL, svcID string) string {
+	return fmt.Sprintf("%s_%s", systemURL, svcID)
+}
```
These two funcs (at least) are free functions, that is, they don't need to access `m` at all. Can we keep them here private but disassociate from the `m Manager` parameter?

PhilipGough replied:
right, yep sure

unleashed commented on pkg/threescale/authorizer/authorizer.go:
```diff
@@ -0,0 +1,276 @@
+
+func (m Manager) setValueFromConfig(systemURL string, request SystemRequest, value *cache.Value) *cache.Value {
+	value.SetRefreshCallback(m.refreshCallback(systemURL, request, m.systemCache.config.numRetryFailedRefresh))
+	value.SetExpiry(time.Now().Add(time.Second * m.systemCache.config.ttlSeconds))
```
I am trying to figure out here if this `m.systemCache.config.ttlSeconds` is any different than the initial value we use to build the cache. Does that ever change? If not it would seem that this ttl is just the same ttl stored in the cache.

PhilipGough replied:
yes thats correct really, i just added it here to make the behaviour explicit but I can remove that if you think its best?

unleashed replied:
Yes, it confused me quite a bit, because my natural reaction was: we keep two ttls because they vary and somehow we must keep the original. Same thing for any other things we might be duplicating without a need because they don't vary.

PhilipGough replied:
gotcha

unleashed reviewed (commented):
Looks ok, with a few comments. The only thing I have doubts about is whether when performing requests it would be worth it to keep the system client as configured (ie. with the URL and token), and whether we can also keep the backend client for the next call and avoid creating it again (that is, with the backend URL but not the exact request to be made, since that varies).

PhilipGough commented on cmd/server/main.go:
```diff
@@ -142,7 +153,9 @@ func cacheConfigBuilder() *threescale.ProxyConfigCache {
-	return threescale.NewProxyConfigCache(cacheTTL, cacheRefreshInterval, cacheUpdateRetries, cacheEntriesMax)
+	ttl := time.Duration(cacheTTL) * time.Second
+	config := authorizer.NewSystemCacheConfig(cacheUpdateRetries, time.Duration(cacheRefreshInterval)*time.Second, ttl, make(chan struct{}))
+	return authorizer.NewSystemCache(cache.NewConfigCache(ttl, cacheEntriesMax), *config)
```
Yep, that's reasonable enough I think

PhilipGough commented:
> Looks ok, with a few comments. The only thing I have doubts about is whether when performing requests it would be worth it to keep the system client as configured (ie. with the URL and token), and whether we can also keep the backend client for the next call and avoid creating it again (that is, with the backend URL but not the exact request to be made, since that varies).

Right so your are considering caching the clients themselves? I guess we could do that (perhaps in another pr?) but I'm wondering how much of an overhead there is from creating and throwing away as opposed to storing and managing?

unleashed commented:
Depends on usage - lots of overhead if the HTTP connections have to be set up from zero for each request and we are contacting a given server frequently enough (ie. we'd want to keep the connection alive).

PhilipGough commented:
@unleashed right that makes sense. Will likely want to reuse the http client then at least. Will resolve that part

philipgough committed:
authorizer: Rework authorizer layout

Removes funcs from Manager which do not directly use any Manager fields.

Expose fields on SystemCacheConfig and embed it directly in SystemCache.
Move defaults to construction of SystemCache and provide in-memory cache
as default.

philipgough committed:
test:authoririzer: Fixes unit test post refactor cache structs

unleashed commented on pkg/threescale/authorizer/authorizer_test.go:
```diff
@@ -236,7 +236,7 @@ func TestManager_CacheRefreshCallback(t *testing.T) {
 
 	m := Manager{}
-	cacheKey := m.generateSystemCacheKey(systemURL, svcID)
+	cacheKey := generateSystemCacheKey(systemURL, svcID)
```
Just a note here that there is a typo in this commit's title.

unleashed reviewed (approved):
Changes look good. Review commits (there is one with a typo) and good to go!

philipgough committed:
dep: Update to latest revision of 3scale backend client

philipgough committed:
authorizer: Changes for latest backend client

philipgough committed:
threescale: Bug fix, service ID not read from correct config

unleashed commented:
Better late than never, Codecov. Good bot!
